# 错误处理

## try-catch 语句

ECMA-262 第 3 版引入了 try-catch 语句，作为 JavaScript 中处理异常的一种标准方式。与 Java 中的 try-catch 语句完全相同。

``` js
try {
    window.someNonexistentFunction();
} catch (error) {
    console.log(error.message);
}
```

### finally 子句

和 Java 一样。无论 try 或 catch 语句块中包含什么代码——甚至 return 语句，都不会阻止 finally 子句的执行。

``` js
try {
    return 2;
} catch (error) {
    return 1;
} finally {
    return 0;
}
```

**以上代码返回 0。**

如果提供 finally 子句，则 catch 子句就成了可选的。

### 错误类型

ECMA-262 定义了 7 种错误类型：

* Error（基类型）
* EvalError（在使用 eval() 函数而发生异常时被抛出。简单的说，如果没有把 eval() 当成函数调用，就会抛出错误）
* RangeError（数值超出相应范围时触发）
* ReferenceError（找不到对象触发）
* SyntaxError
* TypeError（访问不存在的变量，或者变量中保存着意外的类型时）
* URIError（使用 encodeURI() 或 decodeURI()，而 URI 格式不正确时，结果就会导致 URIError）。

最常发生 TypeError 的情况，就是传递给函数的参数事先未经检查，结果传入类型与预期类型不相符。

### 合理使用 try-catch

## 抛出错误
